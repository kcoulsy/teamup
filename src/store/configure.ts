import { configureStore } from '@reduxjs/toolkit';

// Import Reducers
import authReducer from '../reducers/auth';
import teamReducer from '../reducers/team';

const store = configureStore({
    reducer: {
        // reducers here
        auth: authReducer,
        team: teamReducer,
    },
});

export type RootState = ReturnType<typeof store.getState>;

export default store;
