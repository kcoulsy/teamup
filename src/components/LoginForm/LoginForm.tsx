import React from 'react';
import { useHistory } from 'react-router-dom';
import { connect } from 'react-redux';

import { startLogin } from '../../actions/auth';
import { RootState } from '../../store/configure';
import { Card, Form, Input, Button } from 'antd';
import { PATH_REGISTER } from './../../constants/pageRoutes';

interface LoginFormProps {
  startLogin: Function;
  attemptingLogin: boolean;
  loginAttemptFailed: boolean;
}

export const LoginForm: React.FunctionComponent<LoginFormProps> = ({
  startLogin,
  attemptingLogin,
  loginAttemptFailed,
}) => {
  const history = useHistory();
  return (
    <Card title='Login' className='login-form__card' size='small'>
      <Form
        onFinish={({ username, password }) => startLogin(username, password)}>
        <Form.Item
          label='Username'
          name='username'
          rules={[
            {
              required: true,
              message: 'Please input your username!',
            },
          ]}>
          <Input />
        </Form.Item>

        <Form.Item
          label='Password'
          name='password'
          rules={[
            {
              required: true,
              message: 'Please input your password!',
            },
          ]}>
          <Input.Password />
        </Form.Item>
        {loginAttemptFailed ? (
          <Form.Item>Unable to login with those credentials</Form.Item>
        ) : null}
        <Form.Item>
          <Button type='primary' htmlType='submit' loading={attemptingLogin}>
            Login
          </Button>
          <Button
            type='link'
            htmlType='button'
            onClick={() => history.push(PATH_REGISTER)}>
            Not registered? Click here
          </Button>
        </Form.Item>
      </Form>
    </Card>
  );
};

export const mapStateToProps = (state: RootState) => {
  return {
    attemptingLogin: state.auth.attemptingLogin,
    loginAttemptFailed: state.auth.loginAttemptFailed,
  };
};

const mapDispatchToProps = { startLogin };

export default connect(mapStateToProps, mapDispatchToProps)(LoginForm);
