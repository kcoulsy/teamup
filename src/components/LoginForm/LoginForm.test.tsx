import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

import { describe, expect, test, vi } from 'vitest';
import { LoginForm, mapStateToProps } from './LoginForm';
import { RootState } from '../../store/configure';

describe('render login form component', () => {
  test('should render LoginForm component', () => {
    const startLogin = vi.fn();
    const wrapper = shallow(
      <LoginForm
        startLogin={startLogin}
        attemptingLogin={false}
        loginAttemptFailed={false}
      />
    );
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  // TODO - update to react testing library
  // test('should call startLogin on submit', () => {
  //   const startLogin = vi.fn();
  //   const wrapper = shallow(
  //     <LoginForm
  //       startLogin={startLogin}
  //       attemptingLogin={false}
  //       loginAttemptFailed={false}
  //     />
  //   );
  //   const testUser = {
  //     username: 'username',
  //     password: 'password',
  //   };
  //   wrapper
  //     .find('input[name="username"]')
  //     .simulate('change', { target: { value: testUser.username } });
  //   wrapper
  //     .find('input[name="password"]')
  //     .simulate('change', { target: { value: testUser.password } });
  //   wrapper
  //     .find('button[type="submit"]')
  //     .simulate('click', { preventDefault: vi.fn() });

  //   expect(startLogin.mock.calls.length).toBe(1);
  //   expect(startLogin.mock.calls[0][0]).toBe(testUser.username);
  //   expect(startLogin.mock.calls[0][1]).toBe(testUser.password);
  //   expect(shallowToJson(wrapper)).toMatchSnapshot();
  // });

  // TODO - update to react testing library
  // test('should show errors if field is empty', () => {
  //   const startLogin = vi.fn();
  //   const wrapper = shallow(
  //     <LoginForm
  //       startLogin={startLogin}
  //       attemptingLogin={false}
  //       loginAttemptFailed={false}
  //     />
  //   );
  //   const testUser = {
  //     username: 'username',
  //     password: 'password',
  //   };

  //   //testing if password is missing, show the password error
  //   wrapper
  //     .find('input[name="username"]')
  //     .simulate('change', { target: { value: testUser.username } });
  //   wrapper
  //     .find('input[name="password"]')
  //     .simulate('change', { target: { value: '' } });
  //   wrapper
  //     .find('button[type="submit"]')
  //     .simulate('click', { preventDefault: vi.fn() });

  //   expect(startLogin.mock.calls.length).toBe(0);
  //   expect(wrapper.find('.error.message').length).toBe(1);
  //   expect(
  //     wrapper.find('.error.message').contains('Please enter a password')
  //   ).toBe(true);

  //   // typing in that password should remove the error on click
  //   wrapper
  //     .find('input[name="password"]')
  //     .simulate('change', { target: { value: testUser.password } });
  //   wrapper
  //     .find('button[type="submit"]')
  //     .simulate('click', { preventDefault: vi.fn() });

  //   expect(startLogin.mock.calls.length).toBe(1);
  //   expect(wrapper.find('.error.message').length).toBe(0);

  //   //////////////////////

  //   // Now testing if username is missing, show the username error
  //   wrapper
  //     .find('input[name="username"]')
  //     .simulate('change', { target: { value: '' } });
  //   wrapper
  //     .find('input[name="password"]')
  //     .simulate('change', { target: { value: testUser.password } });
  //   wrapper
  //     .find('button[type="submit"]')
  //     .simulate('click', { preventDefault: vi.fn() });

  //   expect(startLogin.mock.calls.length).toBe(1); // still 1 from testing password
  //   expect(wrapper.find('.error.message').length).toBe(1);
  //   expect(wrapper.find('form').hasClass('error')).toBe(true);
  //   expect(
  //     wrapper.find('.error.message').contains('Please enter a username')
  //   ).toBe(true);

  //   // typing in that username should remove the error on click
  //   wrapper
  //     .find('input[name="username"]')
  //     .simulate('change', { target: { value: testUser.username } });
  //   wrapper
  //     .find('button[type="submit"]')
  //     .simulate('click', { preventDefault: vi.fn() });

  //   expect(startLogin.mock.calls.length).toBe(2);
  //   expect(wrapper.find('.error.message').length).toBe(0);
  // });

  // TODO - update to react testing library
  // test('should show error login attempt failed', () => {
  //   const startLogin = vi.fn();
  //   const wrapper = shallow(
  //     <LoginForm
  //       startLogin={startLogin}
  //       attemptingLogin={false}
  //       loginAttemptFailed={false}
  //     />
  //   );

  //   expect(wrapper.find('.error.message').length).toBe(0);

  //   wrapper.setProps({ loginAttemptFailed: true });
  //   expect(wrapper.find('.error.message').length).toBe(1);
  //   expect(wrapper.find('.error.message').contains('Login Invalid')).toBe(true);
  // });

  test('should correctly map state to props', () => {
    const mockState: Partial<RootState> = {
      auth: {
        attemptingLogin: false,
        token: null,
        loginAttemptFailed: false,
        appInitialising: false,
        appInitialised: false,
        attemptingRegister: false,
      },
    };

    // @ts-ignore
    expect(mapStateToProps(mockState).attemptingLogin).toEqual(
      mockState.auth?.attemptingLogin
    );
    // @ts-ignore
    expect(mapStateToProps(mockState).loginAttemptFailed).toEqual(
      mockState.auth?.loginAttemptFailed
    );
  });
});
