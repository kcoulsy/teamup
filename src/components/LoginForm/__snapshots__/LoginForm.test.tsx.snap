// Vitest Snapshot v1

exports[`render login form component > should render LoginForm component 1`] = `
<Card
  className="login-form__card"
  size="small"
  title="Login"
>
  <ForwardRef(InternalForm)
    onFinish={[Function]}
  >
    <FormItem
      label="Username"
      name="username"
      rules={
        [
          {
            "message": "Please input your username!",
            "required": true,
          },
        ]
      }
    >
      <Input
        type="text"
      />
    </FormItem>
    <FormItem
      label="Password"
      name="password"
      rules={
        [
          {
            "message": "Please input your password!",
            "required": true,
          },
        ]
      }
    >
      <Password
        action="click"
        iconRender={[Function]}
        visibilityToggle={true}
      />
    </FormItem>
    <FormItem>
      <Button
        block={false}
        ghost={false}
        htmlType="submit"
        loading={false}
        type="primary"
      >
        Login
      </Button>
      <Button
        block={false}
        ghost={false}
        htmlType="button"
        loading={false}
        onClick={[Function]}
        type="link"
      >
        Not registered? Click here
      </Button>
    </FormItem>
  </ForwardRef(InternalForm)>
</Card>
`;

exports[`render login form component should call startLogin on submit 1`] = `
<Card
  centered={true}
  title="Login"
>
  <Loader
    isLoading={false}
    loadingText="Logging in"
  >
    <form
      className="ui form"
    >
      <div
        className="field"
      >
        <label>
          Username
        </label>
        <input
          name="username"
          onChange={[Function]}
          placeholder="Username"
          type="text"
        />
      </div>
      <div
        className="field"
      >
        <label>
          Password
        </label>
        <input
          name="password"
          onChange={[Function]}
          placeholder="Password"
          type="password"
        />
      </div>
      <button
        className="ui button"
        onClick={[Function]}
        type="submit"
      >
        Login
      </button>
      Not registered? 
      <Link
        to="/register"
      >
        Register
      </Link>
    </form>
  </Loader>
</Card>
`;

exports[`render login form component should render LoginForm component 1`] = `
<Card
  centered={true}
  title="Login"
>
  <Loader
    isLoading={false}
    loadingText="Logging in"
  >
    <form
      className="ui form"
    >
      <div
        className="field"
      >
        <label>
          Username
        </label>
        <input
          name="username"
          onChange={[Function]}
          placeholder="Username"
          type="text"
        />
      </div>
      <div
        className="field"
      >
        <label>
          Password
        </label>
        <input
          name="password"
          onChange={[Function]}
          placeholder="Password"
          type="password"
        />
      </div>
      <button
        className="ui button"
        onClick={[Function]}
        type="submit"
      >
        Login
      </button>
      Not registered? 
      <Link
        to="/register"
      >
        Register
      </Link>
    </form>
  </Loader>
</Card>
`;
